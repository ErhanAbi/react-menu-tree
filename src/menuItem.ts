export interface MenuItem {
  title: string;
  url: string;
  children?: MenuItem[];
}

export const filterMenuItems = (menuItems: MenuItem[], keyword: string) => {
  if (!keyword) {
    return menuItems;
  }

  const filtered: MenuItem[] = menuItems.filter(menuItem => {
    const containsKeyword =
      menuItem.title.toLowerCase().indexOf(keyword.toLowerCase()) > -1;
    const childContainsKeyword: boolean = !!menuItem.children
      ? filterMenuItems(menuItem.children, keyword).length > 0
      : false;

    return containsKeyword || childContainsKeyword;
  });

  return filtered;
};
