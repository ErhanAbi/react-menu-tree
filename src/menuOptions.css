.menu-options {
  margin: 0.75em;
  padding-left: 0.25em;

  line-height: 1.25;
  list-style: none;
}

.menu-options a {
  display: block;
  border-radius: 4px;
  padding: 0.5em 0.75em;

  color: #4a4a4a;
  text-decoration: none;
}

.menu-options a.selected {
  background-color: #3273dc;
  color: #fff;
}

.menu-options a:not(.selected):hover {
  background-color: #f5f5f5;
}

