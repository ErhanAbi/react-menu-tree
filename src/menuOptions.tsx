import * as React from "react";
import { MenuItem } from "./menuItem";
import { trim } from "./trim";
import { HighlightedMenuOption } from "./highlightedMenuOption.";

import styles from "./menuOptions.css";

export const MenuOptions = ({
  options,
  searchedValue,
  selectedOption,
  highlightClass
}: {
  options: MenuItem[];
  searchedValue: string;
  selectedOption: string;
  highlightClass: string;
}) => {
  return (
    <ul className={styles["menu-options"]}>
      {options.map(menuOption => {
        return (
          <li key={menuOption.title}>
            <a
              className={
                trim(selectedOption, "/") === trim(menuOption.url, "/")
                  ? styles.selected
                  : ""
              }
              href={`/${menuOption.url}`}
            >
              <HighlightedMenuOption
                text={menuOption.title}
                searchedValue={searchedValue}
                highlightClass={highlightClass}
              />
            </a>
            {menuOption.children instanceof Array &&
              menuOption.children.length > 0 && (
                <MenuOptions
                  options={menuOption.children}
                  searchedValue={searchedValue}
                  selectedOption={selectedOption}
                  highlightClass={highlightClass}
                />
              )}
          </li>
        );
      })}
    </ul>
  );
};
