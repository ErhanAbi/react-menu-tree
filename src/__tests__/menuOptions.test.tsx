import React from "react";
import renderer from "react-test-renderer";

import { MenuOptions } from "../menuOptions";
import { MenuItem } from "../menuItem";

const testMenuList: MenuItem[] = [
  {
    title: "Getting started",
    url: "getting-started",
    children: [
      { title: "First", url: "getting-started/first" },
      { title: "Installation", url: "getting-started/installation" },
      { title: "Site preparation", url: "getting-started/site-preparation" },
      {
        title: "Creating your site",
        url: "getting-started/creating-your-site"
      },
      {
        title: "Publishing your site",
        url: "getting-started/publishing-your-site"
      }
    ]
  },
  {
    title: "Guides",
    url: "guides",
    children: [
      {
        title: "Adding a Blog",
        url: "guides/adding-a-blog",
        children: [{ title: "First", url: "guides/adding-a-blog/first" }]
      },
      {
        title: "Custom Pages",
        url: "guides/custom-pages",
        children: [{ title: "First", url: "guides/custom-pages/first" }]
      },
      { title: "Enabling Search", url: "guides/enabling-search" },
      {
        title: "Navigation and Sidebars",
        url: "guides/navigation-and-sidebars"
      },
      {
        title: "Translations & Localization",
        url: "guides/trans-and-nav"
      },
      { title: "Versioning", url: "guides/versioning" }
    ]
  }
];

describe("MenuOptions", () => {
  it("correctly renders", () => {
    const wrapper = renderer
      .create(
        <MenuOptions
          options={testMenuList}
          searchedValue=""
          selectedOption=""
          highlightClass=""
        />
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
