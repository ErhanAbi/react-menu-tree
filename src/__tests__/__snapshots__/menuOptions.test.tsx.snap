// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuOptions correctly renders 1`] = `
<ul
  className={undefined}
>
  <li>
    <a
      className=""
      href="/getting-started"
    >
      Getting started
    </a>
    <ul
      className={undefined}
    >
      <li>
        <a
          className=""
          href="/getting-started/first"
        >
          First
        </a>
      </li>
      <li>
        <a
          className=""
          href="/getting-started/installation"
        >
          Installation
        </a>
      </li>
      <li>
        <a
          className=""
          href="/getting-started/site-preparation"
        >
          Site preparation
        </a>
      </li>
      <li>
        <a
          className=""
          href="/getting-started/creating-your-site"
        >
          Creating your site
        </a>
      </li>
      <li>
        <a
          className=""
          href="/getting-started/publishing-your-site"
        >
          Publishing your site
        </a>
      </li>
    </ul>
  </li>
  <li>
    <a
      className=""
      href="/guides"
    >
      Guides
    </a>
    <ul
      className={undefined}
    >
      <li>
        <a
          className=""
          href="/guides/adding-a-blog"
        >
          Adding a Blog
        </a>
        <ul
          className={undefined}
        >
          <li>
            <a
              className=""
              href="/guides/adding-a-blog/first"
            >
              First
            </a>
          </li>
        </ul>
      </li>
      <li>
        <a
          className=""
          href="/guides/custom-pages"
        >
          Custom Pages
        </a>
        <ul
          className={undefined}
        >
          <li>
            <a
              className=""
              href="/guides/custom-pages/first"
            >
              First
            </a>
          </li>
        </ul>
      </li>
      <li>
        <a
          className=""
          href="/guides/enabling-search"
        >
          Enabling Search
        </a>
      </li>
      <li>
        <a
          className=""
          href="/guides/navigation-and-sidebars"
        >
          Navigation and Sidebars
        </a>
      </li>
      <li>
        <a
          className=""
          href="/guides/trans-and-nav"
        >
          Translations & Localization
        </a>
      </li>
      <li>
        <a
          className=""
          href="/guides/versioning"
        >
          Versioning
        </a>
      </li>
    </ul>
  </li>
</ul>
`;
